### PR TITLE
Make the elastic sink a linked transport.

### LINKED ISSUE #681: Pass Elastic errors as events, instead of logging them
**Describe the problem you are trying to solve**

In order to properly react to errors coming from requests sent via elastic sink, the overall errors and the bulk-element errors should be sent as error events to the `err` port of the sink. Currently bulk element errors are only logged at the `ERROR` level, which might quickly fill up disks. Also with error events with a proper format, users are going to be able to handle failures according to their business logic (retry policies while maintaining idempotency etc. etc.)

**Describe the solution you'd like**

Send all sink errors as Error events. This includes bulk element errors. These might be sent as separate events or as part of a single event which contains an array of all bulk errors.

**Notes**

If this is an implementation of an [RFC](https://github.com/tremor-rs/tremor-rfcs) provide a URL
to the RFC this enhancement implements.

If this is a major enhancement or contribution an RFC may be required. It is ok to submit an enhancement
first and our core team will assist with major contributions. In general, major contributions should be
discussed with the community before submission.

### PR DESCRIPTION
**Pull request by mfelsche**
Sending error reponses via err port, success responses via out port.

# Pull request

## Description

Changes the `elastic` sink to be a linked transport. When configured with `linked: true`, it will send error events down the `err` port and success events down the `out` port.

## Related

* Related Issues: fixes #681 
* Related [docs PR](https://github.com/tremor-rs/tremor-www-docs/pull/65)

## Checklist

* [x] The RFC, if required, has been submitted and approved
* [x] Any user-facing impact of the changes is reflected in docs.tremor.rs
* [x] The code is tested
* [x] Use of unsafe code is reasoned about in a comment
* [x] Update CHANGELOG.md appropriately, recording any changes, bug fixes or other observable changes in behavior

### CODE REVIEW

File: src/sink/elastic.rs [hunk 1]
```diff
@@ -63,107 +67,339 @@ pub struct Elastic {
...
+                    if m.insert("time", *time).is_err() {
+                        // ALLOW: this is OK
+                        unreachable!()
+                    };
+                    CBAction::Ack
+                }
+                Err(e) => {
+                    // request failed
+                    // TODO update error metric here?
+                    println!("ES request failed: {:?}", e);
```
Comment (Licenser) [inline src/sink/elastic.rs:339]: lets use error so users can filter it

```suggestion
                    error!("ES request failed: {:?}", e);
```

File: src/sink/elastic.rs [hunk 2]
```diff
@@ -198,6 +445,41 @@ impl Elastic {
...
+        match rx.recv().await {
+            Ok((data, port)) => {
+                let response_event = Event {
+                    id: id_gen.next_id(),
+                    data,
+                    ingest_ns: nanotime(),
+                    origin_uri: Some(origin_uri.clone()),
+                    ..Event::default()
+                };
+                eprintln!("Sending response {:?} via {}", &response_event, port);
```
Comment (Licenser) [inline src/sink/elastic.rs:467]: Do we need to priont on every event? If so can we use `error!` to allow filtering / redirecting

Comment (Licenser) [reply src/sink/elastic.rs:467]: same below

**Review by Licenser — CHANGES_REQUESTED** [review]

**Review by Licenser — CHANGES_REQUESTED** [review]
Looks good :D nice work on adding the errors!

File: src/sink/elastic.rs (hunk 2 above)
Comment (mfelsche) [reply src/sink/elastic.rs:467]: yeah, those are stupid printf-debugging leftovers. gonna clean them up soonish

Comment (Licenser) [reply src/sink/elastic.rs:467]: Ja I do the same :D and then fail to see them

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,11 +2,16 @@
 
 ## Latest
 
+### Breaking Changes
+
+* elastic sink config value `endpoints` renamed to `nodes` [#715](https://github.com/tremor-rs/tremor-runtime/pull/715)
```
[the remark this answers is not in the document]
Comment (mfelsche) [reply CHANGELOG.md:7]: Yeah, we can keep it this way until we break everything with `0.10`. But we definitely should change this.

File: src/sink/elastic.rs [hunk 4]
```diff
@@ -63,107 +67,339 @@ pub struct Elastic {
...
+    Value::from(source)
+}
+
+/**
+Format:
+
+Value:
+{
+    "source": {
+        "`event_id`": ...,
```
[the remark this answers is not in the document]
Comment (mfelsche) [reply src/sink/elastic.rs:121]: clippy required me to add those. Gonna work around it somehow

File: src/sink/elastic.rs [hunk 5]
```diff
@@ -63,107 +67,339 @@ pub struct Elastic {
...
+    payload: Value<'static>,
+    origin_uri: Option<&EventOriginUri>,
+) -> Result<LineValue> {
+    let mut meta = Object::with_capacity(1);
+    let mut es_meta = Object::with_capacity(4);
+    es_meta.insert("id".into(), Value::from(item.id().to_string()));
+    es_meta.insert("index".into(), Value::from(item.index().to_string()));
+    es_meta.insert("doc_type".into(), Value::from(item.ty().to_string()));
+    es_meta.insert(
+        "_version".into(),
```
[the remark this answers is not in the document]
Comment (mfelsche) [reply src/sink/elastic.rs:194]: Good catch, omg!

File: CHANGELOG.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Licenser) [reply CHANGELOG.md:7]: Ja, I generally agree ti's an sensible change just feels like the wrong time for it :D

**Review by Licenser — DISMISSED** [review]
LGTM :+1: great work!

**Review by Licenser — APPROVED** [review]
:+1: woooh!

**Pull request merged**